### ISSUE TITLE
Tick escaping and select narrowing

### ISSUE THREAD

**Issue by EliLichtblau** [schummar/schummar-translate#23]
## Tick escaping 
I'm not fully sure of the specification on tick escaping - I couldn't find any information on it beyond the [docs](https://unicode-org.github.io/icu/userguide/format_parse/messages/#quotingescaping) which states:

* Starting with ICU 4.8, an ASCII apostrophe only starts quoted text if it immediately precedes a character that requires quoting (that is, “only where needed”), and works the same in nested messages as on the top level of the pattern. The new behavior is otherwise compatible; for details see the MessageFormat and MessagePattern (new in ICU 4.8) API docs

Reading the parser implementation in formatjs we see this excerpt
```ts
/**
   * Starting with ICU 4.8, an ASCII apostrophe only starts quoted text if it immediately precedes
   * a character that requires quoting (that is, "only where needed"), and works the same in
   * nested messages as on the top level of the pattern. The new behavior is otherwise compatible.
   */
  private tryParseQuote(parentArgType: ArgType): string | null {
    if (this.isEOF() || this.char() !== 39 /* `'` */) {
      return null
    }

    // Parse escaped char following the apostrophe, or early return if there is no escaped char.
    // Check if is valid escaped character
    switch (this.peek()) {
      case 39 /* `'` */:
        // double quote, should return as a single quote.
        this.bump()
        this.bump()
        return "'"
      // '{', '<', '>', '}'
      case 123:
      case 60:
      case 62:
      case 125:
        break
      case 35: // '#'
        if (parentArgType === 'plural' || parentArgType === 'selectordinal') {
          break
        }
        return null
      default:
        return null
    }
```

Which basically says '{, '<, '}, '< all get escaped away to either EOF or succeeding tick. There is also a special case for "#" in the implementation but this is a tad more work to implement and I haven't investigated the behavior yet.

To handle tick escaping I think while naive a valid approach would be to take the passed ICU string and first pass it through the following.

```ts
type EscapeLike = `'${'{' | '}' | '<' |'>' }`
type EscapeTicks<T extends string> = T extends `${infer Start}${EscapeLike}${string}'${infer End}` ?
                                        `${Start}${EscapeTicks<End>}` : 
                                     T extends `${infer Start}${EscapeLike}${string}` ? Start : T
```

I use [this](https://format-message.github.io/icu-message-format-for-translators/editor.html) website to test with - if this is a different enough version of ICU let me know but I think its valid. 

Here's a [playground link](https://www.typescriptlang.org/play?ssl=43&ssc=72&pln=43&pc=78#code/C4TwDgpgBAaghgGwK7QLxQM7AE4EsB2A5lAD5T5IC2ARhNqVNQPZMIRz4MAicwEA3AFgAUCNCQoAdQAWuPhjBwAxmigBydQzUAdYBrI78+9duxqho4QHoAVDagAlCJSYA3aGzgATAsQ5eoYDhcBF8oAHdZeUUVKBsrMXBoABU8SgAeZIA+KHRkqAgADz58LwwoAAMAEgBvGTkIBWUIAF9aggAzOkdG4BaKkSgoAH4oVNwMpywswagALjGC4ohS8uqazu6pvtr66Ob+2dHxyd6Z4SGF-KKSsswcXyOx2YX8CHdsCxFbeydgJGw+HKHCgcGw2DgICgTA6gSYYCgbHcCEYCCYSgA1uV4okJAAxAheABCaMxGEyNxyeRuS1ua1qWDwRBaNXa+C69AcuEI0j6FSgVisHUJUGF2CwUBqTyc3hJ6Ix6TUagANI5ubzVUqcjcVncANqbehyzGqw1jYIIAC6syGoz1xoxqoAdC6CaUHeTkhaspaBVYoNh2AFqKSMaDSqKRW9iqj5aKmPQgiEbfNyO86C8oHrLfw-eQmJLVYH-oCCpQwKAA40kAhgCJvnYoG7g6Hyh0EwVlNJAkgwGwCvgcFCccJxCle2xmx7MlTFjrVlXvEx8AgoQb2d0CeLgM6XWbtr67S6nVPW+kt9Nd07khOIKf5eTtj7U9mvtZGzKAkmUUhByFAtI0BKACgaDqu0KQJwIZxrg5RKGiGAQF4To9CWnDAIBsaYlASjLiUwDhgEgYSjCgQWnECSjkkPSyqG6QOrSurlIyviql6-7zncLFEKqXAQBW3accxDxELO7EokJlRshyUAADIAIwtG0Gwbpyin8oKkYRmKEotE8CmMQu6xmrJABMLLSVs5kaf6uCwmAgYqEhSGMFCNSqhhsEQSswJplgLnQZiKajJ+DrpOsDrKWZLIVKq6wOOZykOOpcW1HxAktAA1BUOSaRhKw4XhBAoJg7DYEo3Z4DyBGkZ55RShcQypulGGGXcFSZZZnK9P0eYQAgiFQHZUBePxrU5AADMFNHEnREWhlFin9KqyUrWceb5Zwwr4LBgEBG8-ktvKqq4X+FDQIhYIVQG6o1bCOkEXpjWXFm83yoty2OPJvqaf1g3FgCW01iigVhv4pVXd27b0NGdaNQsr55vgBYtEWEBoWWFbYFCxE1nDDb2AACmCg2g2jGOuGCuBwCGjQRHI3ZjsCEaOQCGC4MiULbftvQBa29OeZwI5jlAxPihAYUOrODGScZqlQAAcnAlCtMqXVNgmlC8Kj6vbIcjWjDUWYANJDZwJzpErKs+gs8B4DTbDJEkmRpOemu8FkOQtFAABkYy3mLiFhTefZ3oS073mS6RPp7mYy8sRnq1bqvq3i7t8k8Rt6qbBBjK7yc27AVMOxATuQC7Exu9gWvAJ7UBPS9Wc5+brtS5atuICVLRvj8osk9AoPxvQ7DXf8ocDkOFG4uOoeBxLdHZLkc4J3cgZLiua5mheO5QMe+69Ieffi2F285H7IdsHPYVPqmv4Ysj4T4G+It29TtNlxAmRtcJTKEGJ3-qAoDQOgxg1B9gBIgUBiE2BKGAAmHw+BIEjHIFQWg9ArgALUF4XgEBQHAAmLg5BPA+CpngMgAQ9YqISAAKIYCUHASAslcAYlUBUNQtQ1A1FAS0UBipSBqCyBofWItaH0MgMkXAUdrgrx-r4f+st1YAGUgjYB2DUURDCIBMJYcpbihAeHq2oaUXqwwUxNXMRYyxVjrFNXWMosEaiNHiMkVidIRivBZF6gsMxNjfF+KGNIukUkVIyXsao5STitHMNaAyES+j+SjDCQRK4lDe4AGFEDAQQDg5m0JqAACsICwMCNRcIDNQQIBRJTe2tNmZfkArgRMSRyi53ylAQgHMCoAEk0kAFUh413uL-KewgihgATAREWABxdGPTekAEFsCEE9AAvRDA15eGXOBPR2YxIB37mFWYgSmKLk2RvIZvhszIIvmHd0Z5F4LEjq47IIgsjP2onPOZX9JJ6NnDM4AczFnLLcXQzREipGx0sCLFWGAMBwEIKoAARHVUESyqArAIhgaQTAazBmgI0JQigXKcLBPojQNQiiwWABgFoiLp7QgQF4GAdB2bLiXv8wFSzyQwrhQi84mkhgAD1TFUOgIoYAFUXLoE+X09IPL4UQH5f6IVIr6XI2AJErwABZXgkqMCSAgNQdmJDpX9y+YiwMhAaxgnUFgHBKtBySlJbSpV5jhWUJFl0jAmqvWnQ6EgdmRAl4yt6ekC1EArXZPoJw9Qag4Auu+MqlVlCgA) with the Escape code on top of your extract ICU file. 

If you would be interested in accepting something like this I would be interested in getting the implementation accurate to "#" as well. (I think this is correct for everything else).

## Select Narrowing

This is more of an implementation decision arguably but I would really appreciate select narrowing. Essentially given the ICU
{gender, select, female {you picked female} male {you picked male} other {you selected other}}

The values gender could meaningfully be are "male", "female", ".\*" where .\* is literally anything else. There isn't really a convenient way of expression this in typescript because of union reductions but for me I prefer the type signature 
```ts
type GenderType = "male" | "female" | "other"
```
on consumption you can any cast anything to "other" but this would have to be explicit - this wouldn't allow for typos or something like "males" which I think is preferable. The naive way of doing this is just to slightly adapt your FindBlock, ReadBlock code to emit the type. Again if this is something you would be interested in allowing to be merged I would spend time implementing this in a more sensible way but here is a [playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAaghgGwK7QLxQM7AE4EsB2A5lAD5T5IC2ARhNqVNQPZMIRz4MAicwEA3AFgAUCNCQoAdQAWuPhjBwAxmigBydQzUAdYBrI78+9duxqho4QHoAVDagAlCJSYA3aGzgATAsQ5eoYDhcBF8oAHdZeUUVKBsrMXBoABU8SgAeZIA+KHRkqAgADz58LwwoAAMAEgBvGTkIBWUIAF9aggAzOkdG4BaKkSgoAH4oVNwMpywswagALjGC4ohS8uqazu6pvtr66Ob+2dHxyd6Z4SGF-KKSsswcXyOx2YX8CHdsCxFbeydgJGw+HKHCgcGw2DgICgTA6gSYYCgbHcCEYCCYSgA1uV4okJAAxAheABCaMxGEyNxyeRuS1ua1qWDwRBaNXa+C69AcuEI0j6FSgVisHUJUGF2CwUBqTyc3hJ6Ix6TUagANI5ubzVUqcjcVncANqbehyzGqw1jYIIAC6syGoz1xoxqoAdC6CaUHeTkhaspaBVYoNh2AFqKSMaDSqKRW9iqj5aKmPQgiEbfNyO86C8oHrLfw-eQmJLVYH-oCCpQwKAA40kAhgCJvnYoG7g6Hyh0EwVlNJAkgwGwCvgcFCccJxCle2xmx7MlTFjrVlXvEx8AgoQb2d0CeLgM6XWbtr67S6nVPW+kt9Nd07khOIKf5eTtj7U9mvtZGzKAkmUUhByFAtI0BKACgaDqu0KQJwIZxrg5RKGiGAQF4To9CWnDAIBsaYlASjLiUwDhgEgYSjCgQWnECSjkkPSyqG6QOrSurlIyviql6-7zncLFEKqXAQBW3accxDxELO7EokJlRshyUAADIAIwtG0Gwbpyin8oKkYRmKEotE8CmMQu6xmrJABMLLSVs5kaf6uCwmAgYqEhSGMFCNSqhhsEQSswJplgLnQZiKajJ+DrpOsDrKWZLIVKq6wOOZykOOpcW1HxAktAA1BUOSaRhKw4XhBAoJg7DYEo3Z4DyBGkZ55RShcQypulGGGXcFSZZZnK9P0eYQAgiFQHZUBePxrU5AADMFNHEnREWhlFin9KqyUrWceb5Zwwr4LBgEBG8-ktvKqq4X+FDQIhYIVQG6o1bCOkEXpjWXFm83yoty2OPJvqaf1g3FgCW01iigVhv4pVXd27b0NGdaNQsr55vgBYtEWEBoWWFbYFCxE1nDDb2AACmCg2g2jGOuGCuBwCGjQRHI3ZjsCEaOQCGC4MiULbftvQBa29OeZwI5jlAxPihAYUOrODGScZqlQAAcnAlCtMqXVNgmlC8Kj6vbIcjWjDUWYANJDZwJzpErKs+gs8B4DTbDJEkmRpOemu8Fka3TFALRQAAZGMt5i4hYU3n2d6EtO95kukT5ZOcL0y8sRnq1bqvq3i7t8k8Rt6qbBBjK7ac27AVMOxATuQC7Exu9gWvAPHPuZrn+fm67UuWrbiAlS0b4-KLJPQKD8b0Ow13-OHA5DhRuLjuHwcS3R2S5HOyd3IGS4rmuZoXjuUDHvuvSHgP4thbvOQB2HbAL2FT6pr+GLI+E+BviLdvU7TlcQJkbXCUyhBewIpJbihBUAACIwFiV-uoCgNA6DGDUH2AEiAEGITYEoYACYfD4BQSMcgVBaD0AWCma4a9yhqC8LwCACDgATGoXgngfBUzbGgRAvB8BkDQCuNAtQaCIAYI0KMAA4ujAAyv1fhwA7ax3Wl3Thb5Z5QBEcAcR6CpFggpMUMSNJZbq1khADofRWQqRklyaqvVBTc0jNuSUTxlGqMkXbAAEv1SA2BFQqjVNVTUahtRkKzGaB0pp5biWtM9PBFt9GGJyCQexEiMHSPErlf0G8jrYXBlY2GWEwzQzIsmcJrx0zYEzG8D4uZNLI0LFWDGzgKw42rLWRRcS1HONcXQeioZoEgLYuRYBIkAFQBaoJfxICxK9P8XLGSCklK63UnmKxD0m4GzkvJaBkzujRWMfuayeZhoOX4RAZywY3IeVkOUeEPlQR+T4GkjE01mmOLBC4hAbjwq1EirUTZsUpI1ASjM35KUflDKyjlDagFOCnVoedCG5VKq3WhLCOqtiCmDLGsMuklROomK2D1GyBQBrQGGqNASUBJr3LEfE9R2BnmvLepiD6n1Vo9G9nlcFWldouQOjc7JJ0irQsurCm61UEXWN0imBGdKMQfW+clH6-o-rQABqWDowNsmERhddXJsNMyIwqSjcmgNMbT1xo0yw-cADCiBgIICoczaE1AABWkjAjUXCAzUECAUSU3trTZmX5AK4ETEkcoBd8pQEIBzAqABJc1ABVEe9d7j-xnsIIoYAEwERFsomNsaACC2BCCei6f0hgqTlzgRAdmMSQdB5hVmKQjFZat5Jt8NmCJt5o5YhnKmTtnpzhZFftRHNABZRoGA4CEFUGAmok7Sh0FVHwjBqouha37DUEATAkBQDALgTELkV2IFaFAVd0B12bu3bujELkT2+yYPlegZ6t2Lu5XewC2AlJgPrFRCQCh+F2VwEhBxGCV7Zrjfmwt6QR1jonRAc4mkmoAD1hgiCAA) to an implementation

Thank you so much for this library its really incredible!

**Comment by schummar** [comment]
Thanks for the hints and detailed explanations.

# Regarding tick escaping
Looks promising, but I need to look into it some more. At the moment the library does not recognize any kind of escape sequences. It would be useful, but probably quite complex to catch all cases.

At least escape sequences a rare in translations (in my experience), so it's not too limiting to use `t.unknown` as a workaround in those few cases.

If you want to make a PR, you're welcome to do so. 😉 

# Regarding select narrowing
That one I have immediately built, and it's included in v1.12.0 - it was fun. 😁 The implementation is similar to yours but reuses the existing code a bit better and gives more precise types.

It requires nested values only where needed. So for `{ value, select, option1 {text1} option2 {text2} }`, instead of `{ value: "option1" | "option2" } & { nested: Value }` it calculates `({ value: "option1" } & { nested: Value }) | { value: "option2" }`

And about the "other" problem - there is actually a commonly used workaround. Returning a union of `"option1" | "option2" | (string & {})` will prevent the union being reduced to `string`. That way, code completion will still suggest the explicit options but any string is allowed. The only downside is that it's kind of a hack, so future versions of Typescript might break it.

**Comment by schummar** [comment]
I also already included your escape code. You are right, it is not 100% complete because of the `#` case, but that is one very edgy edge case. Having escapes work in all other cases is still a big win and the `#` would have shown an error before as well, so no downside really.

To finally solve `#` within plural blocks, one would have to integrate the escape logic more with the block parser. Just stripping them in advance will not work, because the behavior depends on the parent block, so let's see...
